Favicon goodies

ryankeairns opened the pull request:
Fixes #65585
^ We'll need to spin off a separate issue to consider allowing custom (user uploaded) favicons.

## Summary

To better differentiate Kibana browser tabs from other Elastic proporties, this PR changes the favicon in the following ways:
1. For production mode, displays the monochrome version of the Elastic cluster [1]
2. For development mode [2a], displays a monochrome hammer icon in the style of the cluster [2b]
3. Changes the favicon fill color based upon the current operating system color theme via the `prefers-color-scheme` media query [3]
4. Removes unnecessary favicon files for unsupported browsers [4]

### Browser screenshots

#### Chrome
<img width="600" alt="Chrome" src="https://user-images.githubusercontent.com/446285/102916076-93075f00-4448-11eb-89f0-7b06a2facf20.png">

#### Firefox
<img width="600" alt="Firefox" src="https://user-images.githubusercontent.com/446285/102916126-aa464c80-4448-11eb-9ce1-fc251ea37c34.png">

#### Edge
<img width="600" alt="Edge" src="https://user-images.githubusercontent.com/446285/102916150-b500e180-4448-11eb-8a12-7dbd6e8ca3bb.png">

#### Safari
<img width="600" alt="Safari" src="https://user-images.githubusercontent.com/446285/102916164-ba5e2c00-4448-11eb-9774-81a813172870.png">

### How to test
1. Start up Kibana; see new favicon in your browser tab.
2. Change system theme, then refresh browser.
3. Change environment mode (I'm actually not sure how to do this, so I've only flipped the tertiary check. Is there a `yarn start` flag or something similar to alter this setting locally?)

### Checklist

Delete any items that are not applicable to this PR.

- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

### Footnotes
[1] The monochrome version is included in the [updated branding guidelines](https://brand.elastic.co/302f66895/p/06c73c-elastic-logo/t/14994f)
[2a] As I understand it, a production instance can be run in dev mode and vice-versa. Regardless, I think it's beneficial to show the development favicon version based upon the mode as it is a clear, simple indicator of which mode Kibana is running under.
[2b] Since this is only for dev mode, it feels like an opportunity to have a little fun with the favicon. Feedback welcome; let me know of any potential concerns you might have in this regard.
[3] There a few ways to accomplish this effect, but the majority favored the [SVG](https://css-tricks.com/svg-favicons-and-all-the-fun-things-we-can-do-with-them/#dark-mode-support) [style](https://flaviocopes.com/dark-mode-favicon/) [approach](https://catalin.red/svg-favicon-light-dark-theme/) versus [using Javascript](https://css-tricks.com/how-to-create-a-favicon-that-changes-automatically/). After reading through those and testing some things out, I chose the SVG with embedded `style` approach as it avoids the need for Javascript and reduces the number of graphic files needed. Also, it just plain worked with our existing setup.
[4] The major browsers, supported by Kibana, [support SVG favicons with the exception of Safari](https://caniuse.com/link-icon-svg). A fallback, black PNG favicon has been included to accommodate Safari (see screenshot above) and, as a side effect, likely works for several other browsers. For more information on favicon files, [this article was very helpful](https://css-tricks.com/favicon-quiz/).
<img width="310" alt="Screen Shot 2020-12-22 at 9 09 14 AM" src="https://user-images.githubusercontent.com/446285/102913853-fd1e0500-4444-11eb-9cbe-92f71094c197.png">
 

cc:/ @alexfrancoeur @elastic/kibana-design

ryankeairns commented on x-pack/plugins/security/server/authorization/reset_session_page.tsx:
```diff
@@ -46,33 +46,12 @@ export function ResetSessionPage({
-          href={`${uiPublicUrl}/favicons/safari-pinned-tab.svg`}
-        />
-        <link rel="shortcut icon" href={`${uiPublicUrl}/favicons/favicon.ico`} />
+        {/* The alternate icon is a fallback for Safari which does not yet support SVG favicons */}
```
This reset session page does not have the full complement of environment variables available to it, so I have excluded the environment check/favicon switch in favor of always showing the monochrome cluster version.

ryankeairns replied:
@jportner thanks for the review and offer to assist. If you still have the time, I'd appreciate your contribution to this branch. Some time has passed since you commented, so I understand if you are busy with other things.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/)
* Commit: 11d728f23ef284a34cc4cde227a31c6bf2fd8ad6
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Integration Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/execution/node/327/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/","number":"96016","commit":"11d728f23ef284a34cc4cde227a31c6bf2fd8ad6"}]}
PIPELINE-->

alexfrancoeur commented:
@ryankeairns glad to see we're addressing these concerns! And I like the idea for identifying tabs that are in development mode. Curious, have we synced with Cloud on these changes at all? If I recall, having the same icon was part of the unified header work.  cc: @VijayDoshi for visibility

Random thought. But because we are using an SVG now, I wonder if we can tie into some of the custom header / footer work (https://github.com/elastic/kibana/issues/17298), with space specific or global settings to persist a banner. I'm not sure how much I like the idea, if it's appropriate with our branding guidelines or what the UX could be, but I could see admins starting to define spaces or Kibana instances by color. A yellow persisted banner means QA/Dev environments while green means prod. Those color definitions could theoretically extend to the logo itself. It's probably not worth investing time into this idea now, but if we see this use case being adopted, maybe we can revisit.

ryankeairns commented:
> @ryankeairns glad to see we're addressing these concerns! And I like the idea for identifying tabs that are in development mode. Curious, have we synced with Cloud on these changes at all? If I recall, having the same icon was part of the unified header work. cc: @VijayDoshi for visibility

I have not caught up with Cloud, yet, but will do so after the holiday break (cc: @gjones ). At a minimum, this separates us (Kibana and likely the Cloud UI) from the docs, marketing sites, discuss, etc.

> Random thought. But because we are using an SVG now, I wonder if we can tie into some of the custom header / footer work (#17298), with space specific or global settings to persist a banner. I'm not sure how much I like the idea, if it's appropriate with our branding guidelines or what the UX could be, but I could see admins starting to define spaces or Kibana instances by color. A yellow persisted banner means QA/Dev environments while green means prod. Those color definitions could theoretically extend to the logo itself. It's probably not worth investing time into this idea now, but if we see this use case being adopted, maybe we can revisit.

Yep, this does open us up to future enhancements with an outcome similar to the github-style favicons.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/)
* Commit: 91301b4818ffe3aba1c8e7f04239408f94c49942

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/d744eaec8ce40516fd0d6e81ba784e2dec5ca24f) | [after](https://github.com/elastic/kibana/commit/91301b4818ffe3aba1c8e7f04239408f94c49942) | diff |
| --- | --- | --- | --- |
| `default` | 47142 | 47896 | +754 |
| `oss` | 27711 | 27705 | -6 |

### History
* :broken_heart: [Build #96016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/) failed 11d728f23ef284a34cc4cde227a31c6bf2fd8ad6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/","number":"96016","commit":"11d728f23ef284a34cc4cde227a31c6bf2fd8ad6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/","number":"96033","commit":"91301b4818ffe3aba1c8e7f04239408f94c49942"}]}
PIPELINE-->

jportner reviewed (approved):
Tested reset session page and it renders as expected with the favicon. LGTM, one optional comment below.

willemdh commented:
@ryankeairns Love this PR, got at least 20+ open tabs to a Development (basic), a QA, a PR (platinum) and a dedicated monitoring cluster. Please consider allowing us to set a custom icon per cluster type.

gjones commented:
> I have not caught up with Cloud, yet, but will do so after the holiday break (cc: @gjones ). At a minimum, this separates us (Kibana and likely the Cloud UI) from the docs, marketing sites, discuss, etc.

From the above it sounds like the proposal is for Cloud to also use the monochrome cluster for prod and the hammer for development? This makes sense to me, as Dream Machine is a Cloud based service, it makes sense for them to follow suit, I'll start communications with that team.

gjones commented:
Rereading the development part, so the hammer is for kibana development mode rather than for use to differentiate local environments or staging/qa from prod?

ryankeairns commented:
> > I have not caught up with Cloud, yet, but will do so after the holiday break (cc: @gjones ). At a minimum, this separates us (Kibana and likely the Cloud UI) from the docs, marketing sites, discuss, etc.
> 
> From the above it sounds like the proposal is for Cloud to also use the monochrome cluster for prod and the hammer for development? This makes sense to me. As Dream Machine is a Cloud based service, it seems to make sense for them to follow suit, I'll start communications with that team.

Yes, that is what I'm suggesting. Your note about Dream Machine also makes sense.

> Rereading the development part, so the hammer is for kibana development mode rather than for use to differentiate local environments or staging/qa from prod?

To directly answer your question, yes, that is what this PR does. I would consider this PR as a first iteration on improving the favicon situation. The environment variable I have used here was readily available and provided a quick win. 

More can be done - such as allowing for custom, user-defined favicons as the comment above suggest - and I presume we could get into further granularity with a localhost variation, status indicators (as GitHub uses), and more. In fact, these two specific points may even be related in the sense that custom favicons could be defined per environment (not only for cloud but self-managed situations where the number of type of environments varies widely).

Let me know if I'm misreading your comment... it may well be I'm overlooking something with this approach.

pgayvallet commented:
> More can be done - such as allowing for custom, user-defined favicons as the commenter above suggests

This would be covered by https://github.com/elastic/kibana/issues/75377 afaik.

pgayvallet reviewed (approved):
LGTM (technically) for core changes

ryankeairns commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/)
* Commit: 60edfa56a9a49008c7164d0125289a12f12f130d
* This comment will update when the build is complete

### Failed CI Steps
* [X-Pack Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/execution/node/329/log)

### History
* :green_heart: [Build #96033](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/) succeeded 91301b4818ffe3aba1c8e7f04239408f94c49942
* :broken_heart: [Build #96016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/) failed 11d728f23ef284a34cc4cde227a31c6bf2fd8ad6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/","number":"96016","commit":"11d728f23ef284a34cc4cde227a31c6bf2fd8ad6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/","number":"96033","commit":"91301b4818ffe3aba1c8e7f04239408f94c49942"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/","number":"96675","commit":"60edfa56a9a49008c7164d0125289a12f12f130d"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96697/)
* Commit: 41d05be03311e52c280dab145110faf049b478ee

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/5e5a4102043c3d4501904982ecd59811970ccbc0) | [after](https://github.com/elastic/kibana/commit/41d05be03311e52c280dab145110faf049b478ee) | diff |
| --- | --- | --- | --- |
| `default` | 47266 | 48020 | +754 |
| `oss` | 27713 | 27707 | -6 |

### History
* :broken_heart: [Build #96675](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/) failed 60edfa56a9a49008c7164d0125289a12f12f130d
* :green_heart: [Build #96033](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/) succeeded 91301b4818ffe3aba1c8e7f04239408f94c49942
* :broken_heart: [Build #96016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/) failed 11d728f23ef284a34cc4cde227a31c6bf2fd8ad6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/","number":"96016","commit":"11d728f23ef284a34cc4cde227a31c6bf2fd8ad6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/","number":"96033","commit":"91301b4818ffe3aba1c8e7f04239408f94c49942"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/","number":"96675","commit":"60edfa56a9a49008c7164d0125289a12f12f130d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96697/","number":"96697","commit":"41d05be03311e52c280dab145110faf049b478ee"}]}
PIPELINE-->

jportner commented:
Added e4634fa to address https://github.com/elastic/kibana/pull/86804#discussion_r551067880!

<details><summary>Click for screenshots</summary>

Dev mode (`yarn start`)

![image](https://user-images.githubusercontent.com/5295965/103577031-0e81fb00-4ea2-11eb-997e-2e929e4ebc96.png)

Prod mode (`node scripts/kibana -c config/kibana.dev.yml  --dist`)

![image](https://user-images.githubusercontent.com/5295965/103577243-6caede00-4ea2-11eb-91ea-69d2114aacdf.png)

</details>

ryankeairns commented:
After receiving feedback from @spalger , I will be closing this PR and taking a different approach. He, and the Kibana team historically speaking, advise against leveraging the env mode value in such manner as I've done here. Instead, he proposed that we take a _built_ vs _running from source_ approach where the favicon can be swapped during the build process.

In a new PR, I'll default to the alternative, likely-not-to-remain-a-hammer favicon for the _running from source_ case and update the build script to swap in the monochrome cluster favicon. This also means we only ship the cluster favicon version in our builds.

@jportner thank you for the review and commit. My apologies that we are now taking a different approach.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96763/)
* Commit: e4634fa7bbda2f84be1691d027511793ffe0dd15

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/5e5a4102043c3d4501904982ecd59811970ccbc0) | [after](https://github.com/elastic/kibana/commit/e4634fa7bbda2f84be1691d027511793ffe0dd15) | diff |
| --- | --- | --- | --- |
| `default` | 47266 | 48020 | +754 |
| `oss` | 27713 | 27707 | -6 |

### History
* :green_heart: [Build #96697](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96697/) succeeded 41d05be03311e52c280dab145110faf049b478ee
* :broken_heart: [Build #96675](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/) failed 60edfa56a9a49008c7164d0125289a12f12f130d
* :green_heart: [Build #96033](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/) succeeded 91301b4818ffe3aba1c8e7f04239408f94c49942
* :broken_heart: [Build #96016](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/) failed 11d728f23ef284a34cc4cde227a31c6bf2fd8ad6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96016/","number":"96016","commit":"11d728f23ef284a34cc4cde227a31c6bf2fd8ad6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96033/","number":"96033","commit":"91301b4818ffe3aba1c8e7f04239408f94c49942"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96675/","number":"96675","commit":"60edfa56a9a49008c7164d0125289a12f12f130d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96697/","number":"96697","commit":"41d05be03311e52c280dab145110faf049b478ee"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/96763/","number":"96763","commit":"e4634fa7bbda2f84be1691d027511793ffe0dd15"}]}
PIPELINE-->

ryankeairns commented:
Replaced by #87271
